### PR TITLE
Override replaceAll and sort on List implementations.

### PR DESCRIPTION
**Pull request by donraab**
Signed-off-by: Donald Raab <Donald.Raab@bnymellon.com>

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java [hunk 1]
```diff
@@ -180,6 +182,17 @@ public T set(int index, T element)
...
 
+    /**
+     * Overrides default method from List.
+     *
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        Arrays.sort(this.items, 0, this.size(), comparator);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: I was expecting sort() would delegate to sortThis(). I was expecting an attempt to add the implementation to MutableList. It's possible that would lead to diamond inheritance problems and need to get pushed down, but it probably still makes sense to delegate from one to the other. Thoughts? In this case, the implementation of sort() and sortThis() in ArrayAdapter are different and probably shouldn't be. Either they should both get the optimization for small sizes, or neither.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java [hunk 2]
```diff
@@ -205,6 +218,15 @@ public T set(int index, T element)
         return result;
     }
 
+    /**
+     * @since 10.0
+     */
+    @Override
+    public void replaceAll(UnaryOperator<T> operator)
+    {
+        InternalArrayIterate.replaceAll(this.items, this.size(), operator);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: It seems unlikely that this is the right place for this implementation, simply because you had to add an import for InternalArrayIterate. This should be pulled up to whichever superclass already imports InternalArrayIterate.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java (hunk 1 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: Alternately, sortThis could delegate to sort(). Either way, it should appear that there are no new implementations of sorting.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java [hunk 3]
```diff
@@ -429,28 +430,40 @@ public void toSortedListBy()
...
-    @Test
+    @Test(expected = UnsupportedOperationException.class)
     public void set()
     {
-        Verify.assertThrows(UnsupportedOperationException.class, () -> this.classUnderTest().castToList().set(0, 1));
+        this.classUnderTest().castToList().set(0, 1);
     }
 
-    @Test
+    @Test(expected = UnsupportedOperationException.class)
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java]: I would prefer to keep using Verify.assertThrows() in favor of @Test(expected = class). It works better with multiple assertions in a method, and it works better when delegating to super.sameTestMethod().

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java [hunk 4]
```diff
@@ -429,28 +430,40 @@ public void toSortedListBy()
...
-        Verify.assertThrows(UnsupportedOperationException.class, () -> this.classUnderTest().castToList().addAll(0, Lists.fixedSize.of()));
+        this.classUnderTest().castToList().addAll(0, Lists.fixedSize.of());
+    }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void sort()
+    {
+        this.classUnderTest().castToList().sort(Comparator.naturalOrder());
+    }
+
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java]: When we have methods named sort() and sortThis(), as we probably do in this JUnit test, we should probably keep them together.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java [hunk 5]
```diff
@@ -1132,4 +1134,20 @@ public void forEachWithIndex()
...
+        Assert.assertEquals(Lists.mutable.with(2, 4, 6, 8), integers);
+    }
+
+    @Test
+    public void sort()
+    {
+        MutableList<Integer> integers = this.newWith(1, 2, 3, 4);
+        integers.sort(Comparator.reverseOrder());
+        Assert.assertEquals(Lists.mutable.with(4, 3, 2, 1), integers);
+    }
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java]: I think the thread-safety tests might be missing.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java (hunk 1 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: I tried this initially, but ran into issues with subList().sort().  This method caused StackOverflows errors.

```
protected void defaultSort(Comparator<? super T> comparator)
{
    Collections.sort(this, comparator);
}
```
I just tried this approach again, and remembered why I opted to go with explicit implementations.  I'll see if there is something to do that will fix this.  Help appreciated.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java (hunk 2 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: Fair point.  I moved it up to the abstract class.

Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: Strike the last comment.  Moving this up and removing the method causes duplication of the code in FastList as well.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java (hunk 1 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/ArrayAdapter.java]: I'm going to try and implement sort and delegate to sort from sortThis.  I think this conceptually makes the most sense.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java (hunk 5 above)
Comment (nikhilnanivadekar) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java]: Do the thread-safety tests need to be in Scala?

**Comment by motlin** [comment]
Historically they were but they don’t need to be.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java [hunk 6]
```diff
@@ -262,18 +263,33 @@ public int getBatchCount(int batchSize)
...
+
+    /**
+     * @since 10.0
+     */
     @Override
     public FastList<T> sortThis(Comparator<? super T> comparator)
     {
-        Arrays.sort(this.items, 0, this.size, comparator);
-        return this;
+        return (FastList<T>) super.sortThis(comparator);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: these overrides are because we return FastList. Should we break compatibility and replace it with a MutableList? It will be safe at runtime if users cast. We can then replace it by the default method.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/ListAdapter.java [hunk 7]
```diff
@@ -226,17 +224,31 @@ public void forEach(int fromIndex, int toIndex, Procedure<? super T> procedure)
...
+
+    /**
+     * @since 10.0
+     */
     @Override
     public ListAdapter<T> sortThis(Comparator<? super T> comparator)
     {
-        Iterate.sortThis(this.delegate, comparator);
-        return this;
+        return (ListAdapter<T>) super.sortThis(comparator);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/ListAdapter.java]: Same comment as FastList.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/RandomAccessListAdapter.java [hunk 8]
```diff
@@ -252,17 +250,22 @@ public void forEach(int fromIndex, int toIndex, Procedure<? super T> procedure)
...
-        Iterate.sortThis(this.delegate, comparator);
-        return this;
+        return (RandomAccessListAdapter<T>) super.sortThis(comparator);
     }
 
     @Override
     public RandomAccessListAdapter<T> sortThis()
     {
-        return this.sortThis(Comparators.naturalOrder());
+        return (RandomAccessListAdapter<T>) super.sortThis();
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/RandomAccessListAdapter.java]: Same comment as FastList

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java [hunk 9]
```diff
@@ -1658,15 +1656,7 @@ private ArrayListIterate()
      */
     public static <T> ArrayList<T> sortThis(ArrayList<T> list, Comparator<? super T> comparator)
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java]: seems like we do not have replaceAll() on ArrayListIterate. Do we want to add it as a separate commit?

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java [hunk 10]
```diff
@@ -205,6 +206,18 @@ public void removeFromIndex()
         Verify.assertThrows(UnsupportedOperationException.class, () -> this.unmodifiableList.remove(0));
     }
 
+    @Test
+    public void replaceAll()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.unmodifiableList.replaceAll(e -> e));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java]: I prefer having Verify.assertThrows() but if rest of the tests are @ throws, then these ones should also be consistent.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java [hunk 11]
```diff
@@ -168,13 +168,23 @@
...
+    }
 
     /**
      * Sorts the internal data structure of this list and returns the list itself as a convenience.
+     * @since 10.0
      */
-    MutableList<T> sortThis();
+    default MutableList<T> sortThis()
+    {
+        this.sort(null);
```
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java]: I'm not sure it's necessary.  I believe this produces the same result.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java [hunk 12]
```diff
@@ -168,13 +168,23 @@
 
     /**
      * Sorts the internal data structure of this list and returns the list itself as a convenience.
+     * @since 10.0
```
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java]: I wasn't sure at first but decided to add it since I added the default implementation that wasn't there before.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java (hunk 5 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/AbstractListTestCase.java]: We should eventually convert them all to Java.  I added them here to be consistent for now as we don't have an alternative yet.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 6 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: I think it's important not to break backwards compatibility here.  I was torn on whether to use the cast or reimplement the method as it is simple enough to do, but I didn't want to duplicate code, and the cast is safe here.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/ListAdapter.java (hunk 7 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/ListAdapter.java]: See above.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/RandomAccessListAdapter.java (hunk 8 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/RandomAccessListAdapter.java]: See above.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java (hunk 9 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java]: What would replaceAll do on ArrayListIterate?  Delegate to ArrayList?

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java (hunk 10 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java]: Not sure what you want me to do here.  Is this inconsistent in this class or with other classes?

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java]: Yes, just verified

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 6 above)
Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: 👍

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java (hunk 9 above)
Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java]: Sorry, I thought it was ArrayListAdapter. My bad.

**Comment by nikhilnanivadekar** [comment]
This looks good to me. @motlin can you please review? Since this is fairly central change more eyes the better. Thanks!

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/QuintupletonList.java [hunk 13]
```diff
@@ -133,6 +136,34 @@ public T set(int index, T element)
...
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        T[] array = (T[]) new Object[]{this.element1, this.element2, this.element3, this.element4, this.element5};
+        Arrays.sort(array, comparator);
+        this.element1 = array[0];
+        this.element2 = array[1];
+        this.element3 = array[2];
+        this.element4 = array[3];
+        this.element5 = array[4];
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/QuintupletonList.java]: Is this significantly different than the default implementation?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java [hunk 14]
```diff
@@ -666,6 +667,24 @@ public T remove(int index)
...
+
+    /**
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> c)
+    {
+        throw new UnsupportedOperationException("Cannot call sort() on " + this.getClass().getSimpleName());
+    }
+
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java]: Where is sortThis()? I was expecting it would already exist in this class. I'm sure it always threw UnsupportedOperationException, but was it throwing an unhelpful message like "Cannot call set()" instead of "Cannot call sortThis()"?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java [hunk 15]
```diff
@@ -608,63 +608,6 @@ public long injectInto(long injectedValue, LongObjectToLongFunction<? super T> f
...
-            }
-        }
-    }
-
-    @Override
-    public MutableList<T> sortThis()
-    {
-        return this.sortThis(Comparators.naturalOrder());
-    }
-
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: Did this stuff move?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java [hunk 16]
```diff
@@ -430,6 +431,25 @@ public int lastIndexOf(Object o)
         return -1;
     }
 
+    /**
+     * @since 10.0
+     */
+    @Override
+    public void replaceAll(UnaryOperator<E> operator)
+    {
+        this.lists.forEachWith(FastList::replaceAll, operator);
```
[the remark this answers is not in the document]
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java]: +1. I would expect `List::replaceAll`

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java [hunk 17]
```diff
@@ -262,18 +263,33 @@ public int getBatchCount(int batchSize)
...
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        if (this.size > 1)
+        {
+            Arrays.sort(this.items, 0, this.size, comparator);
+        }
+    }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: It's not clear to me where this was moved from. The superclass doesn't know it can delegate to the backing array, right? What was the old behavior?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/MultiReaderFastList.java [hunk 18]
```diff
@@ -917,6 +918,40 @@ public int lastIndexOf(Object o)
...
+        this.acquireWriteLock();
+        try
+        {
+            this.delegate.sort(comparator);
+        }
+        finally
+        {
+            this.unlockWriteLock();
+        }
+    }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/MultiReaderFastList.java]: Is sortThis() overridden? I was expecting sort() to be the method before sortThis().

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableList.java [hunk 19]
```diff
@@ -221,6 +222,24 @@ public void forEachWithIndex(int fromIndex, int toIndex, ObjectIntProcedure<? su
...
+
+    /**
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        throw new UnsupportedOperationException("Cannot call sort() on " + this.getClass().getSimpleName());
+    }
+
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableList.java]: Is sortThis() overridden?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java [hunk 20]
```diff
@@ -1658,15 +1656,7 @@ private ArrayListIterate()
...
-        int size = list.size();
-        if (ArrayListIterate.canAccessInternalArray(list))
-        {
-            Arrays.sort(ArrayListIterate.getInternalArray(list), 0, size, comparator);
-        }
-        else
-        {
-            Collections.sort(list, comparator);
-        }
+        list.sort(comparator);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java]: It's already efficient?

File: scala-unit-tests/src/test/scala/org/eclipse/collections/impl/list/mutable/SynchronizedMutableListScalaTest.scala [hunk 21]
```diff
@@ -28,6 +28,15 @@ class SynchronizedMutableListScalaTest extends SynchronizedMutableCollectionTest
...
 
+    @Test
+    def sort_synchronized
+    {
+        this.assertSynchronized
+        {
+            this.classUnderTest.sort(null)
+        }
+    }
+
```
Comment (motlin) [inline scala-unit-tests/src/test/scala/org/eclipse/collections/impl/list/mutable/SynchronizedMutableListScalaTest.scala]: Is sortThis tested in the same class?

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java [hunk 22]
```diff
@@ -61,6 +62,31 @@ public void testGetClass()
...
+        this.list.shuffleThis();
+        this.list.sort(null);
+        Assert.assertEquals(
+                this.getNStrings(),
+                this.list);
+    }
+
     @Test
     public void sortThis()
     {
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java]: Is sortThis() missing coverage for the override that takes a comparator?

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java (hunk 3 above)
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java]: Still relevant I think.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java (hunk 4 above)
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java]: Still relevant I think.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java (hunk 10 above)
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java]: Other classes should also use assertThrows

**Comment by motlin** [comment]
We've also had trouble in the past with coverage tools when one test
overrides another and one has an expected exception and the other does not.
I don't think JUnit 5 uses the expected mechanism either.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java (hunk 10 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableListTest.java]: Please feel free to fix those other classes.

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java (hunk 3 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableListTestCase.java]: Fixed.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java]: The default implementation was added in 10.0.  Please suggest a longer description you think would be acceptable and I will add it.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MutableList.java]: Good point.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java [hunk 23]
```diff
@@ -355,6 +357,29 @@ public boolean retainAllIterable(Iterable<?> iterable)
         throw new UnsupportedOperationException("Cannot call retainAllIterable() on " + this.getClass().getSimpleName());
     }
 
+    /**
+     * @since 10.0
```
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java]: Fixed.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java]: ```private final FastList<FastList<E>> lists = FastList.newList();```

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java (hunk 22 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java]: I'm looking at overall coverage for this entire hierarchy now.  Stay tuned.

**Comment by nikhilnanivadekar** [comment]
This I think is ready to merge, unless @motlin has any concerns.

**Comment by nikhilnanivadekar** [comment]
Please do not merge this till M2 is done.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java (hunk 14 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java]: Do our tests check the message, or just the exception type?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java (hunk 15 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: Why is this stuff gone?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java]: Do you get a compiler error when you change it to `List::replaceAll`?

[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java]: I don't get a compiler error now.  This is not a CompositeList.  It is a CompositeFastList.  I can probably replace with a higher interface.  I can keep it with the concrete implementation as I have it.  I'm not convinced there is a benefit either way, so I'm not compelled to change it.  What do you see as the benefit of the change?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java (hunk 14 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java]: I have to admit to not understanding this question at first.  The method `sortThis()` does not exist in `AbstractImmutableList` because it does not extend `MutableList`.  As you know, our `ImmutableList` implementations have to implement `List` and `sort` is a mutating method on `List` so it should throw `UnsupportedOperationException`.  It was previously throwing an unhelpful message as you point out which is why I provided an override.

Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java]: The tests just check the exception type, not the message.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java (hunk 15 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: I removed it.  There have been improvements to the `sort` implementation in the JDK with the addition of `TimSort`.  I don't think this code pulls its weight since we started compiling with Java 8.  Now that I am overriding sort in all of the `AbstractMemoryEfficientMutableList` subclasses, this code is even less beneficial.

If you think it is a mistake this remove this optimization, let's discuss how we should add it back.  It should probably move from `sortThis` to `sort`.

Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: I'm looking to move this code down to `AbstractMemoryEfficientMutableList` removing the size check and deleting all of the current specialized implementations in the subclasses.  Make sense?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableList.java [hunk 24]
```diff
@@ -124,6 +127,33 @@ public void clear()
...
 
+    /**
+     * This method checks if comparator is null and use a ComparableComparator if it is.
+     *
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        this.insertionSort(Comparators.comparableComparatorIfNull(comparator));
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableList.java]: Do you have a JMH test for this? The extra virtual method calls from using a comparator may hurt performance.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java (hunk 14 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/AbstractImmutableList.java]: Right, I forgot that the interfaces are different here.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java (hunk 15 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: Not trying to encourage a particular optimization, quite the opposite. I find this stuff hard to review. The most important issue is that without overriding sort() and replaceAll(), behavior is wrong on thread-safe collections. So I came into this expecting additional overrides.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java [hunk 25]
```diff
@@ -262,18 +263,33 @@ public int getBatchCount(int batchSize)
...
 
+    /**
+     * Overrides default method from List.
+     *
+     * @since 10.0
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        if (this.size > 1)
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: ComparableTimSort already has this optimization.

```java
        int nRemaining  = hi - lo;
        if (nRemaining < 2)
            return;  // Arrays of size 0 and 1 are always sorted
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/MultiReaderFastList.java [hunk 26]
```diff
@@ -1419,6 +1454,15 @@ public void forEachWithIndex(int fromIndex, int toIndex, ObjectIntProcedure<? su
...
+         * @since 10.0
+         */
+        @Override
+        public void sort(Comparator<? super T> comparator)
+        {
+            this.getDelegate().sort(comparator);
+        }
+
         @Override
         public MutableList<T> sortThis()
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/MultiReaderFastList.java]: I see sort and sortThis together in the inner class. They should be together in the outer class too.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/SynchronizedMutableList.java [hunk 27]
```diff
@@ -216,6 +217,30 @@ public T get(int index)
...
+     */
+    @Override
+    public void sort(Comparator<? super T> comparator)
+    {
+        synchronized (this.getLock())
+        {
+            this.getDelegate().sort(comparator);
+        }
+    }
+
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/SynchronizedMutableList.java]: Keep sort() with sortThis() on line 440?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableList.java (hunk 19 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/UnmodifiableMutableList.java]: Same here line 183

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java (hunk 20 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ArrayListIterate.java]: Should this be marked deprecated and inlineable?

File: unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java [hunk 28]
```diff
@@ -227,8 +266,116 @@ public void aggregateByNonMutating()
...
         Function<String, String> groupBy = Functions.getStringPassThru();
         Function2<Integer, String, Integer> sumAggregator = (aggregate, value) -> aggregate + Integer.parseInt(value);
-        MapIterable<String, Integer> actual = this.classUnderTest().aggregateBy(groupBy, () -> 0, sumAggregator);
-        MapIterable<String, Integer> expected = FastList.newList(this.classUnderTest()).aggregateBy(groupBy, () -> 0, sumAggregator);
+        MapIterable<String, Integer> actual =
+                this.classUnderTest().aggregateBy(groupBy, () -> 0, sumAggregator);
+        MapIterable<String, Integer> expected =
+                FastList.newList(this.classUnderTest()).aggregateBy(groupBy, () -> 0, sumAggregator);
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java]: A few unrelated changes here?

Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableListTestCase.java]: Yes.  A general rule I like to apply.  Don't make me scroll.

File: scala-unit-tests/src/test/scala/org/eclipse/collections/impl/list/mutable/SynchronizedMutableListScalaTest.scala (hunk 21 above)
Comment (donraab) [reply scala-unit-tests/src/test/scala/org/eclipse/collections/impl/list/mutable/SynchronizedMutableListScalaTest.scala]: Good point.  Added tests for both forms of sortThis.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableList.java (hunk 24 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractMemoryEfficientMutableList.java]: No. This code is only for FixedSizedLists of size 2 to 6.  I can try and measure it if you think it is important.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/factory/Comparators.java [hunk 29]
```diff
@@ -444,4 +461,15 @@ public int compare(Pair<?, T> p1, Pair<?, T> p2)
...
+    private static class ComparableComparator<T> implements SerializableComparator<T>
+    {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public int compare(T one, T two)
+        {
+            return ((Comparable<T>) one).compareTo(two);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/block/factory/Comparators.java]: There are actually more.  I went searching to see if one existed already.  There is a private ComparableComparator I found in `javax.swing.table.TableRowSorter`.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 25 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: Nice.  I will remove the optimization then.
